util: Refactor logging code into a global object

jimpo opened the pull request:
This is purely a refactor with no behavior changes.

This creates a new class `BCLog::Logger` to encapsulate all global logging configuration and state.

Commit:
util: Establish global logger object.

The object encapsulates logging configuration, and in a later commit,
set up routines will also be moved into the class.

Commit:
util: Move debug file management functions into Logger.

Commit:
util: Encapsulate logCategories within BCLog::Logger.

Commit:
util: Refactor GetLogCategory.

Changing parameter types from pointers to references and uint32_t to
BCLog::LogFlags simplies calling code.

laanwj commented:
Concept ACK, splitting the logging functionality out of util makes sense, util.cpp is large, haphazard and this is a clearly distinguishable concern from the rest.

MarcoFalke commented:
Concept ACK. Needs rebase

sipa commented on src/util.h:
```diff
@@ -75,99 +62,10 @@ inline std::string _(const char* psz)
-#endif
-
 template<typename... Args>
 bool error(const char* fmt, const Args&... args)
```
I think this should move to logging as well, so the cyclic dependency between logging and util can be avoided.

jimpo replied:
`LogPrintf` is also used in TraceThread, and I don't think either of them actually belong in logging (the whole point of this PR is to create a more cohesive separation). I can very easily imagine another method that gets added to util.h in the future requiring logging as well.

The dependency isn't really cyclic -- `logging.cpp` includes `util.h`, and `util.h` includes `logging.h`, but `logging.h` does not include `util.h`.

sipa replied:
My view is that you should always treat the .h and the .cpp file as one unit. While this may indeed not be a cyclic dependency for the compiler, it is certainly one between the two modules semantically: logging can't work without util, and util can't work without logging. Finding out why that is the case helps creating a cleaner separation.

You're right that util is going to depend on logging though, as it contains a number of higher level functions. However, It looks like logging only really needs util for finding the debug log path, though. All the rest is in utiltime. This seems easy to fix, by instead having init query the path and call a setter on logging for it for example. If other reviewers are fine with that, we can fix that up in another PR.

Empact replied:
Last I looked, when reviewing #13021, I saw another approach to solving the cyclic dependency is to include utiltime.h in logging.cpp and splitting out args and path handling from util.h. Didn’t go all the way to implementing that but would be happy to as a follow-on to #13021.

Empact replied:
Lol, didn’t finish reading your comment @sipa. Anyway I’m for it.

jimpo replied:
Fixed with d903273cf8e51d2789ba8a88c608a66e63dcdc14.

sipa reviewed (commented):
Concept ACK

MarcoFalke commented:
Needs rebase again. (Since this has some Concept ACKs, but is somewhat largish, it might be easier to merge in two steps. First the move-only commit in a separate pull request and then the refactoring in this pull request)

Commit:
scripted-diff: Rename BCLog::Logger member variables.

-BEGIN VERIFY SCRIPT-
sed -i "s/fileout/m_fileout/" src/logging.h src/logging.cpp
sed -i "s/mutexDebugLog/m_file_mutex/" src/logging.h src/logging.cpp
sed -i "s/vMsgsBeforeOpenLog/m_msgs_before_open/" src/logging.h src/logging.cpp
sed -i "s/logCategories/m_categories/" src/logging.h src/logging.cpp
sed -i "s/fPrintToConsole/m_print_to_console/" src/logging.h src/logging.cpp src/init.cpp
sed -i "s/fPrintToDebugLog/m_print_to_file/" src/logging.h src/logging.cpp src/init.cpp src/test/test_bitcoin.cpp src/bench/bench_bitcoin.cpp
sed -i "s/fLogTimestamps/m_log_timestamps/" src/logging.h src/logging.cpp src/init.cpp
sed -i "s/fLogTimeMicros/m_log_time_micros/" src/logging.h src/logging.cpp src/init.cpp
sed -i "s/fReopenDebugLog/m_reopen_file/" src/logging.h src/logging.cpp src/init.cpp
sed -i "s/fStartedNewLine/m_started_new_line/" src/logging.h src/logging.cpp
-END VERIFY SCRIPT-

Commit:
util: Store debug log file path in BCLog::Logger member.

This breaks the cyclic between logging and util.

jnewbery commented on src/logging.h:
```diff
@@ -19,15 +21,7 @@ static const bool DEFAULT_LOGIPS        = false;
-
-extern bool fLogTimestamps;
-extern bool fLogTimeMicros;
 extern bool fLogIPs;
```
In commit _util: Establish global logger object._ : any reason to exclude `fLogIPs` from `g_logger`?

jimpo replied:
It's not actually used by the logger, it's just used in the net code. Arguably it should live there instead.

jnewbery reviewed (commented):
Tested ACK d903273cf8e51d2789ba8a88c608a66e63dcdc14. Couple of comments inline but nothing to prevent merge.

jamesob reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/12954/commits/d903273cf8e51d2789ba8a88c608a66e63dcdc14

- [x] `./src/qt/bitcoin-qt -regtest -conf="$(pwd)/bitcoin.conf" -debug=1`
  - Displays all available logs.
- [x] `./src/qt/bitcoin-qt -testnet -conf="$(pwd)/bitcoin.conf" -debug=0`
  - Displays only uncategorized logs.
- [x] `./src/qt/bitcoin-qt -testnet -conf="$(pwd)/bitcoin.conf" -debug=1`
  - Displays all available logs.
- [x] `./src/qt/bitcoin-qt -testnet -conf="$(pwd)/bitcoin.conf" -help`
  - Displays available logging categories for `-debug`.
- [x] `./src/bitcoind -debug=foobar`
  - Warns of invalid logging category.
- [x] `./src/bitcoind -debug=1 -debugexclude=net,addrman,rand,leveldb`
  - Displays all available logs (confusing, but this matches behavior on master).
- [x] `./src/bitcoind -debugexclude=net,addrman,rand,leveldb`
  - Excludes specified log categories.
- [x] `./src/bitcoind -testnet -config="$(pwd)/bitcoin.conf" -debug=reindex -reindex`
  - Shows reindex-related log messages.
- [x] `./src/bitcoind -testnet -config="$(pwd)/bitcoin.conf" -debug=net -reindex`
  - Shows no reindex-related log messages, but plenty of net-related ones.

Nice change. The one nit I had definitely shouldn't hold this up.

jimpo commented on src/logging.cpp:
```diff
@@ -127,19 +124,17 @@ const CLogCategoryDesc LogCategories[] =
+        flag = BCLog::ALL;
+        return true;
+    }
+    for (unsigned int i = 0; i < ARRAYLEN(LogCategories); i++) {
```
Nice, fixed. Also gets rid of the dependency of logging on utilstrencodings.

jnewbery commented:
utACK 8c2d695c4. Only difference is changing to range based for loops per https://github.com/bitcoin/bitcoin/pull/12954#discussion_r184163863

sipa commented:
utACK 8c2d695c4a45bdd9378c7970b0fcba6e1efc01f9. Nice to see this encapsulated.
